apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -530,297 +533,605 @@ public Flow(long maxByteCount, long maxElementCount, Action<Task> registerTaskFn
                 {
-                    _extendQueue.EnqueueAll(ackIds);
+                    _softStopCts.Token.ThrowIfCancellationRequested();
+                    var msg = msgs[msgIndex];
+                    msgs[msgIndex] = null;
+                    await _taskHelper.ConfigureAwait(_flow.Process(msg.CalculateSize(), async () =>
+                    {
+                            // Running async. Common data needs locking
+                            lock (_lock)
+                        {
+                            _softStopCts.Token.ThrowIfCancellationRequested();
+                            _userHandlerInFlight += 1;
+                        }
+                            // Call user message handler
+                            var reply = await _taskHelper.ConfigureAwaitHideErrors(() => _handlerAsync(msg.Message, _hardStopCts.Token), Reply.Nack);
+                            // Lock msgsIds, this is accessed concurrently here and in HandleExtendLease().
+                            lock (msgIds)
+                        {
+                            msgIds.Remove(msg.AckId);
+                        }
+                            // Lock ack/nack-queues, this is accessed concurrently here and in "master" thread.
+                            lock (_lock)
+                        {
+                            _userHandlerInFlight -= 1;
+                            (reply == Reply.Ack ? _ackQueue : _nackQueue).Enqueue(msg.AckId);
+                        }
+                            // Ids have been added to ack/nack-queue, so trigger a push.
+                            _eventPush.Set();
+                    }));
                 }
-                _qEvent.Set();
             }
 
-            private async Task AutoExtend(HashSet<string> ackIds, CancellationTokenSource allMsgsHandled)
+            private void HandleExtendLease(HashSet<string> msgIds)
             {
-                // Moves ackIds into the extend-queue as required for auto lease extension
-                // This loop is not infinite; it terminates when allMsgsHandledToken is cancelled.
-                try
+                if (_softStopCts.IsCancellationRequested)
                 {
-                    while (true)
+                    // No further lease extensions once stop is requested.
+                    return;
+                }
+                bool anyMsgIds;
+                lock (msgIds)
+                {
+                    anyMsgIds = msgIds.Count > 0;
+                    if (anyMsgIds)
                     {
-                        bool isCancelled = await _taskHelper.ConfigureAwaitHideCancellation(
-                            () => _scheduler.Delay(_autoExtendInterval, allMsgsHandled.Token));
-                        if (isCancelled)
+                        lock (_lock)
                         {
-                            return;
+                            // Only enqueue ack IDs that are not already being acked.
```

Comment:
I don't understand what the comment has to do with the code.

Reply:
Erroneous comment, from previous code iteration. Removed.